polynote-server/src/main/scala/polynote/server/KernelPublisher.scala

```diff
@@ -247,24 +246,36 @@ object KernelPublisher {
     subscriberVersions: ConcurrentHashMap[SubscriberId, (GlobalVersion, Int)])(
     subscriberId: SubscriberId,
     update: NotebookUpdate
-  ): Task[Unit] = versionRef.modify {
-    case (globalVersion, notebook) =>
-      val nextVersion = globalVersion + 1
-      val result = (nextVersion, update.applyTo(notebook))
-      (result, (subscriberId, update.withVersions(nextVersion, update.localVersion)))
-  }.tap(publishUpdates.publish1).flatMap {
-    case (subscriberId, update) => ZIO(versions.add(update.globalVersion, update))
+  ): TaskB[Unit] = {
+    versionRef.access.flatMap {
```

Comment:
@jeremyrsmith we discussed this last week - since `applyTo` could throw, I needed to switch from `Ref#modify` to `Ref#access` and this is how I ended up doing it.